libs/contrib/Control/Algebra/CyclicGroup.idr

```diff
@@ -0,0 +1,60 @@
+import Control.Algebra
+
+import Data.ZZ
+
+%access public export
+%default total
+
+----------------------------------------
+
+-- The Verified constraints here are required
+-- to work around an interface resolution bug.
+
+mtimes' : VerifiedMonoid a => (n : Nat) -> a -> a
+mtimes' Z x = neutral
+mtimes' (S k) x = x <+> mtimes' k x
+
+gtimes : VerifiedGroup a => (n : ZZ) -> a -> a
+gtimes (Pos k) x = mtimes' k x
+gtimes (NegS k) x = mtimes' (S k) (inverse x)
+
+----------------------------------------
+
+mtimesTimes : VerifiedMonoid a => (x : a) -> (n, m : Nat) ->
+  mtimes' (n + m) x = mtimes' n x <+> mtimes' m x
+mtimesTimes x Z m = sym $ monoidNeutralIsNeutralR $ mtimes' m x
+mtimesTimes x (S k) m =
+  rewrite mtimesTimes x k m in
+    semigroupOpIsAssociative x (mtimes' k x) (mtimes' m x)
+
+-- TODO: prove this. It's definitely true in spirit, and I'm pretty
+-- sure the implementation of gtimes is correct, but working with ZZ
+-- is not easy.
+postulate
+gtimesTimes : VerifiedGroup a => (x : a) -> (n, m : ZZ) ->
+  gtimes (n + m) x = gtimes n x <+> gtimes m x
+
+gtimesCommutes : VerifiedGroup a => (x : a) -> (n, m : ZZ) ->
+  gtimes n x <+> gtimes m x = gtimes m x <+> gtimes n x
+gtimesCommutes x n m =
+  rewrite sym $ gtimesTimes x n m in
+    rewrite plusCommutativeZ n m in
+      gtimesTimes x m n
```

Comment:
Isn't this just a special case of commutativity of `ZZ` addition? I mean `plusCommutativeZ (gtimes n x) (gtimes m x)` has exactly this type, doesn't it?

Reply:
No, `gtimes n x` produces an instance of the group type, not a `ZZ`.